### ISSUE TITLE
Moving kubeadm out of k/k repository

### ISSUE THREAD

**Issue by dims** [kubernetes/kubernetes#78534]
Slack discussion: https://kubernetes.slack.com/archives/C13J86Z63/p1556823569185700

Was also discussed in the code-organization sub project of sig-arch.

Note from @timothysc :
```
TL;DR again I won't care if APIs are flushed out and promoted.
Biggest ones are ComponentConfig.
Till then, nope
```

**Comment by dims** [comment]
Note from @neolit123 [4:01 PM]
```
ok, now i remember.
https://github.com/kubernetes/kubelet/tree/master/config/v1beta1
^ if this staging for the kublet then kubeadm can move only after the component config work there is done.
```

**Comment by dims** [comment]
Note from @luxas [4:06 PM]
```
Yes, but we're using the internal types & conversions in `pkg/kubelet/apis/config`
Yeah
So first let's decide where to put the conversions & the "hub" version (internal in 99% of the cases)
From a staging perspective
```

**Comment by neolit123** [comment]
https://github.com/kubernetes/kubernetes/tree/f14b128ea6429a34b83578d4c2a69a99296a8d31/pkg/util/initsystem

is currently a bit of a problem. it's being used *only* by kubeadm, TMK.
https://cs.k8s.io/?q=pkg%2Futil%2Finitsystem&i=nope&files=&repos=

this either has to move to staging or back to the kubeadm package.
my vote is to move it back to kubeadm.

**Comment by liggitt** [comment]
> my vote is to move it back to kubeadm.

Agreed

**Comment by neolit123** [comment]
xref 
https://github.com/kubernetes/kubeadm/issues/1600
https://github.com/kubernetes/kubeadm/issues/1205

**Comment by timothysc** [comment]
Given that this is going to take a while and is a pretty large logistical effort given all the tentacles, would it be ok with this group ( @dims @liggitt ) to open issues in the kubeadm repo and we can triage and address?  This way it's part of our milestone burndown.

**Comment by neolit123** [comment]
i was thinking of having the import boss issue as the parent one and if a certain topic gets complicated (i can think of a couple...) we can log an child issue.
today we did some work on the  easier parts.

technically this k/k issue might be a bit redundant, but it's OK to have it too.

**Comment by dims** [comment]
@timothysc @neolit123 please cross link when you can. Let's keep this open at least for 1.16 cycle.

**Comment by BenTheElder** [comment]
cc

**Comment by neolit123** [comment]
opened the topic about the kubeadm release process and packaging:
https://github.com/kubernetes/kubernetes/issues/83127

**Comment by listx** [comment]
Dumb question: Why are we trying to move kubeadm out of k/k repository? I guess just for better code organization? I'm having trouble finding a clear ELI5 answer in the links. Thanks!

**Comment by neolit123** [comment]
@listx 
the reasons are tribal knowledge but they benefit both the kubernetes repository and kubeadm.

a couple of examples:
- faster development process of kubeadm; merging PRs without flakes and not having to run kube-up tests on a unrelated deployer.
- less noise from kubeadm issues and PRs in k/k.
(users might still report kubeadm issues in k/k, but ideally we should move them to k/kubeadm).

**Comment by neolit123** [comment]
https://github.com/kubernetes/kubernetes/issues/83127#issuecomment-554789034

**Comment by dims** [comment]
cc @carlory @pacoxu

**Comment by carlory** [comment]
According to `.import-restrictions` file in `cmd/kubeadm`, kubeadm is no longer dependent on k/k, so it's time to move forward with this issue. I don't know much about the release and build processes, but I'm willing to help move this forward.

Can you provide some guidance if possible? @pacoxu @neolit123 @dims

**Comment by neolit123** [comment]
@carlory i always supported the idea but that's not a small topic.

https://github.com/kubernetes/kubeadm/issues/1949

basically, this needs a new KEP. last KEP had ~200 comments and got blocked by SIG Release mostly. SIG testing also has interests because kind uses kubeadm.
kubectl was planned to be extracted completely first but that didn't happen.
the main problem of the story is that we don't know how to build and release a component that lives outside of k/k as part of the k/k release artifacts.

personally i would like to:
- see how kubectl moves out first and kubeadm can follow in its steps
- continue versioning kubeadm with the k8s versioning

fundamental questions that i have are:
- what problems are we solving by moving components out / splitting the monolith and do we really need to?

**Comment by BenTheElder** [comment]
> faster development process of kubeadm; merging PRs without flakes and not having to run kube-up tests on a unrelated deployer.

At least with respect this problem we've put a lot of work into de-flaking the CI here.

I think it will be more expensive than helpful to move these components out at this point. It makes releasing and qualification a lot more complex, we're struggling through this right now with cloud provider removal which has been painful despite being more clearly decoupled and giving big wins in isolating vendor specific dependencies and binary bloat (especially as users are unlikely to use more than one of these as a time).

With kubeadm and kubectl ... I don't think either are causing major negative impacts to the rest of the repo or vice versa, at least not in ways we can't mitigate without moving out of the repo.

We have a lot of options for isolating things within the repo (import restrictions, modules / workspace, staging, ..) and we actually have decent tooling to skip running irrelevant tests these days, we just need to tune it. We can add `skip_if_only_changed `^cmd/kubeadm/.*` to truly irrelevant jobs (We should be careful about this to avoid skipping on actually relevant, but we're under utilizing it currently)

https://docs.prow.k8s.io/docs/jobs/#trigger-types

For issue triage nothing is perfect and people file issues in the wrong repo all the time, but we have the labeling and we have the github.com/kubeadm and github.com/kubectl issue trackers.

**Comment by neolit123** [comment]
@BenTheElder i agree there have been a lot of improvements in k/k testing and various processes, and i would like to see an update for the benefits vs complexity for component extraction.

@soltysh @seans3 do you have a summary of why kubectl did not move out of staging and start its own versioning?
https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubectl

**Comment by aojea** [comment]
> what problems are we solving by moving components out / splitting the monolith and do we really need to?

this, I don't see why we need to add new problems on something that is working fine and AFAIK does not have any issue

**Comment by soltysh** [comment]
> @soltysh @seans3 do you have a summary of why kubectl did not move out of staging and start its own versioning?
> https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubectl

We've started talking with sig-release about that possibility, but we never had the capacity to finish the move. So that's the primary reason behind who not :sweat_smile:

**Comment by shanduur** [comment]
> what problems are we solving by moving components out / splitting the monolith and do we really need to?

In case of kubeadm being part of k/k, [the API definitions](http://k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta3) are not importable. You need to do in-place replace (`replace k8s.io/kubernetes => k8s.io/kubernetes v1.30.3`) or manually copy the code. Neither option is good. Maybe moving just the API to https://github.com/kubernetes/kubeadm will be enough?

**Comment by neolit123** [comment]
> > what problems are we solving by moving components out / splitting the monolith and do we really need to?
> 
> In case of kubeadm being part of k/k, [the API definitions](http://k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta3) are not importable. You need to do in-place replace (`replace k8s.io/kubernetes => k8s.io/kubernetes v1.30.3`) or manually copy the code. Neither option is good. Maybe moving just the API to https://github.com/kubernetes/kubeadm will be enough?

IIRC, the problem around that is that the kubeadm API imports some core API bits and then that kubeadm API cannot be imported again in k/k unless it matches exactly the state of core APIs that is in k/k. AFAIK, the solution for that is using staging. but i don't think we want staging for kubeadm bits.

**Comment by shanduur** [comment]
> the problem around that is that the kubeadm API imports some core API bits 

It's no more an issue 😃

```
$ find ./kubernetes/cmd/kubeadm -type f -exec grep -H 'k8s.io/kubernetes' {} \+ | grep -v 'k8s.io/kubernetes/cmd/kubeadm'
./kubernetes/cmd/kubeadm/app/util/staticpod/utils.go:// Copied from k8s.io/kubernetes/pkg/util/hash/hash.go#DeepHashObject
./kubernetes/cmd/kubeadm/.import-restrictions:  # prevent any k8s.io/kubernetes imports outside of this package
```

**Comment by neolit123** [comment]
> > the problem around that is that the kubeadm API imports some core API bits
> 
> It's no longer an issue 😃
> 
> ```
> $ find ./kubernetes/cmd/kubeadm -type f -exec grep -H 'k8s.io/kubernetes' {} \+ | grep -v 'k8s.io/kubernetes/cmd/kubeadm'
> ./kubernetes/cmd/kubeadm/app/util/staticpod/utils.go:// Copied from k8s.io/kubernetes/pkg/util/hash/hash.go#DeepHashObject
> ./kubernetes/cmd/kubeadm/.import-restrictions:  # prevent any k8s.io/kubernetes imports outside of this package
> ```

i mean these imports:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go#L20-L21

**Comment by dims** [comment]
/close

Time to drop my lance

![image](https://github.com/user-attachments/assets/b633186e-74be-43a5-94b9-4146b2ca99b5)
